runtime/lua/vim/iter.lua

```diff
@@ -0,0 +1,770 @@
+--- Iterator implementation.
+
+---@class Iter
+---@field next function Return the next value in the iterator
+---@field _table ?table Underlying table data (table iterators only)
+---@field _head ?number Index to the front of a table iterator (table iterators only)
+---@field _tail ?number Index to the end of a table iterator (table iterators only)
+local Iter = {}
+Iter.__index = Iter
+Iter.__call = function(self)
+  return self:next()
+end
+
+--- Special case implementations for iterators on list-like tables.
+local ListIter = {}
+ListIter.__index = setmetatable(ListIter, Iter)
+ListIter.__call = function(self)
+  return self:next()
+end
+
+--- Special case implementations for iterators on map-like tables.
+local MapIter = {}
+MapIter.__index = setmetatable(MapIter, Iter)
+MapIter.__call = function(self)
+  return self:next()
+end
+
+---@private
+local function unpack(t)
+  if type(t) == 'table' then
+    return _G.unpack(t)
+  end
+  return t
+end
+
+---@private
+local function pack(...)
+  if select('#', ...) > 1 then
+    return { ... }
+  end
+  return ...
+end
+
+--- Add a filter step to the iterator pipeline.
+---
+--- Example:
+--- <pre>lua
+--- local bufs = vim.iter(vim.api.nvim_list_bufs()):filter(vim.api.nvim_buf_is_loaded)
+--- </pre>
+---
+---@param f function(...):bool Takes all values returned from the previous stage in the pipeline and
+---                            returns false or nil if the current iterator element should be
+---                            removed.
+---@return Iter
+function Iter.filter(self, f)
+  local next = self.next
+  self.next = function(this)
+    while true do
+      local args = pack(next(this))
+      if args == nil then
+        break
+      end
+      if f(unpack(args)) then
+        return unpack(args)
+      end
```

Comment:
```suggestion
      local unpacked = unpack(args)
      if f(unpacked) then
        return unpacked
      end
```

Would avoid duplicate `unpack` calls. Same for a couple of the other functions. (Or is there special arg passing semantics with Lua that I'm missing?)

Reply:
The pack/unpack dance is necessary because we don’t know how many args the iterator function might return.

```
local unpacked = unpack(args)
```

will drop all but the first value.

We could avoid duplicate unpacking through a helper function maybe, though I’m not sure if that’s better from a performance perspective (I’m not sure how expensive unpack is versus another function call)